org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/util/InsertionOffsets.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (c) 2014, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.xbase.ui.util;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.nodemodel.ICompositeNode;
+import org.eclipse.xtext.nodemodel.ILeafNode;
+import org.eclipse.xtext.nodemodel.util.NodeModelUtils;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Iterables;
+
+/**
+ * @author Anton Kosyakov - Initial contribution and API
+ */
+public class InsertionOffsets {
+
+	public int before(EObject element) {
+		return NodeModelUtils.findActualNodeFor(element).getOffset();
+	}
+
+	public int after(EObject element) {
+		return NodeModelUtils.findActualNodeFor(element).getEndOffset();
+	}
+
+	public int inEmpty(EObject element) {
+		ICompositeNode compositeNode = NodeModelUtils.findActualNodeFor(element);
+		Optional<ILeafNode> openingBraceNode = Iterables.tryFind(compositeNode.getLeafNodes(), (node) -> "{".equals(node.getText()));
```

Comment:
Not a big fan of these usages of optionals. Checking for `isPresent` is usually not good style. You probably want to do something like `Iterables.tryFind(..).transform(node->getOffset() + 1).or(compositeNode::getEndOffset)`

Reply:
Or preferably a plain for loop.